base/combinatorics.jl

```diff
@@ -242,7 +242,16 @@ function invperm(p::Union{Tuple{},Tuple{Int},Tuple{Int,Int}})
     isperm(p) || throw(ArgumentError("argument is not a permutation"))
     p  # in dimensions 0-2, every permutation is its own inverse
 end
-invperm(a::Tuple) = (invperm([a...])...,)
+
+function invperm(P::NTuple{N,T}) where {N,T}
+    isperm(P) || throw(ArgumentError("argument is not a permutation"))
+    ntuple(Val(N)) do i
+        for j in P
+            @inbounds P[j]==i && return j
+        end
+        return first(P) # DO NOT REMOVE: necessary for type stable code_gen
```

Comment:
You could make this an error instead. Can you avoid the upfront `isperm` test by putting it here instead?

Reply:
doing it right now, it is much faster without `isperm` call